### ISSUE TITLE
Support for refactor/rename of an entity

### ISSUE THREAD

**Issue by tudorache** [protegeproject/webprotege#226]
Currently, it is not possible to change the IRI of an entity. It would be useful to have support for refactoring in the UI.

**Comment by matthewhorridge** [comment]
I've changed the issue title to be a bit more specific.

The IRI may be tied to the display name, so this needs to be taken into consideration.  Also, if you change the IRI prefix settings WebProtege will actually ask you if you want to update existing prefixes.

**Comment by lonniev** [comment]
The issue topic is either misleading or too general; really, the crucial missing capability is the ability to alter the IRI for a specific entity. This is different than the automated refactoring of the IRI base path following a change of the New Entity Setting for the IRIs. The use case is: as an Ontologist, after changing the default IRI setting from using hash code fragments, I want to be able to alter the IRI fragment for a specific entity to use the natural-language (e.g. English) name of the entity and not to use the original hashcode.

(It seems that one might be able to force-alter IRIs by downloading the ontology, hand-editing it (or using Protege), and then uploading the altered file.

**Comment by lonniev** [comment]
Attempting this is a nice opportunity to try out the Download / modify / Upload / merge capability -- and that is not an unpleasant experience.

What happens (of course) is that when the merge is performed, it regards the loss of the original as a Delete of the corresponding entity; the use of the new IRI is regarded as a new entity Create; and -- for some reason -- the deletions are not honored in the merge.

Therefore, the effective merged ontology has both the original hashcode-referenced entity and the new language-referenced entity. The UI then shows the Entity Trees with one element per Display Name. Apparently, the hashcode IRIs win in that battle for appearance. (Guava coders, add a custom Ordering class for your Sets and then discriminate on both Display Name AND IRI rather than just on Display Name. Then the Set.add will preserve uniqueness based on IRI and Names.)

The final cleanup is, in WebProtege, after the merge, to delete the original hashcode IRI entities and then the sought language-named IRI entities will appear in the UI widgets.

Lesson learned: set one's IRI pattern BEFORE adding any entities not much later!

**Comment by matthewhorridge** [comment]
I've just tried this using a simple example, which worked as expected.  Do you have a specific example that I could use to reproduce this behaviour?

**Comment by lonniev** [comment]
Do you mean that you were able to merge in the modified ontology having new
IRIs and the merged result had the original hashed entities deleted?

If so, great! But I don’t know why then I had to clean mine up a bit. It
could be that the original ontology first started with WebProtege default
IRIs (using the stanford site and hash for the base IRI and hashes for the
fragments (the stuff after the # in the IRI). Then, I switched the default
IRI pattern to be a stanford host with “/incose/cea#” as the end of the
path. Then I switched from hash names to entity names. Then I created a few
more entities to see if I liked the resulting IRIs. And only then did I try
to figure out how to convert the original IRIs that were using hash
fragment strings to use the entity name. So then I did the download,
offline replace, upload, merge task.



-- 
Lonnie VanZandt
303-900-3048

On 11 November 2016 at 12:07:37, Matthew Horridge (notifications@github.com)
wrote:

I've just tried this using a simple example, which worked as expected. Do
you have a specific example that I could use to reproduce this behaviour?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/protegeproject/webprotege/issues/226#issuecomment-260031305>,
or mute the thread

**Comment by matthewhorridge** [comment]
> Do you mean that you were able to merge in the modified ontology having new
IRIs and the merged result had the original hashed entities deleted?

Yes that's right.  It was only a small example, but the idea is basically the same no matter what the size.  I'm just wondering if there's something in your particular ontology that triggered a bug.

**Comment by Wolkenstein** [comment]
2020 and this important feature still remains on the wishlist. Please implement it.

**Comment by ivenger** [comment]
Agreed, this and entity deletion garbage collection (#639 and #604) are probably the biggest blockers for adoption for me.
